### PR TITLE
Feature/etl batch driver

### PR DESCRIPTION
**Pull request by rohitsinha54**
The PR was reviewed against etl-batch-driver branch here: https://github.com/caskdata/cdap/pull/1889

Merging it to develop. 
Build running here: http://builds.cask.co/browse/CDAP-DT4-5

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StreamToStructuredRecordTransform.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                                                                 LongWritable, StructuredRecord> {
+  private static final String SCHEMA = "schema";
+  private static final String FORMAT_NAME = "format.name";
+  private static final String FORMAT_SETTING_PREFIX = "format.setting.";
+
+  private static SchemaWrapper schemaWrapper = null;
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(StreamToStructuredRecordTransform.class.getName());
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StreamToStructuredRecordTransform.java]: Need to use class.getSimpleName() instead?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StreamToStructuredRecordTransform.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  private SchemaWrapper getSchemaWrapper(Schema streamBodySchema) throws Exception {
+    // create the new schema with timestamp and headers to reatin them in StructuredRecord
+    List<Schema.Field> fields = Lists.newArrayList(
+      Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+      Schema.Field.of("headers", Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING))));
+    fields.addAll(streamBodySchema.getFields());
+
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    for (Map.Entry<String, String> entry : getContext().getRuntimeArguments().entrySet()) {
+      if (entry.getKey().startsWith(FORMAT_SETTING_PREFIX)) {
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StreamToStructuredRecordTransform.java]: Move this String constant to this method if not being used elsewhere? Thought it was a property to be provided by the user.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StructuredRecordToGenericRecordTransform.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Transform {@link StructuredRecord} to {@link GenericRecord}
+ */
+public class StructuredRecordToGenericRecordTransform extends Transform<LongWritable, StructuredRecord,
+                                                               LongWritable, GenericRecord> {
+  private final Map<Integer, Schema> schemaCache = Maps.newHashMap();
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(StructuredRecordToGenericRecordTransform.class.getName());
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StructuredRecordToGenericRecordTransform.java]: SimpleName instead?

File: pom.xml [hunk 4]
```diff
@@ -141,6 +141,7 @@
   <modules>
     <module>cdap-common-unit-test</module>
     <module>cdap-api</module>
+    <module>cdap-app-templates</module>
```
Comment (gokulavasan) [inline pom.xml]: Why is this added? cdap-app-templates build is triggered only when "templates" profile is enabled. Similar to examples.

**Comment by gokulavasan** [comment]
Few comments. I believe rest of the code was reviewed before. rest LGTM :+1:

File: pom.xml (hunk 4 above)
Comment (rohitsinha54) [inline pom.xml]: I think @sreevatsanraman added this so the package also builds templates. 
I will remove it. Sounds good @sreevatsanraman

**Comment by rohitsinha54** [comment]
New Build: http://builds.cask.co/browse/CDAP-DT4-6

**Comment by gokulavasan** [comment]
LGTM :+1:

**Comment by rohitsinha54** [comment]
Random tests keeps failing. New build: http://builds.cask.co/browse/CDAP-DT4-8

**Comment by rohitsinha54** [comment]
Reopened : https://issues.cask.co/browse/CDAP-535

also opened a jira (CDAP-2040) for metrics here: https://issues.cask.co/browse/CDAP-1683

**Comment by rohitsinha54** [comment]
Test passed here: http://builds.cask.co/browse/CDAP-DT4-9
